packages/babel-plugin-proposal-partial-application/src/index.js

```diff
@@ -0,0 +1,286 @@
+   * and gives them a uniques name.
+   * @param {Object} node
+   * @param {Object} scope
+   * @returns {Array<Expression>} the arguments
+   */
+  function unwrapAllArguments(node, scope) {
+    const clone = t.cloneNode(node);
+    clone.arguments.forEach(argument => {
+      if (argument.type === "ArgumentPlaceholder") {
+        argument.type = "Identifier";
+        argument.name = scope.generateUid("_argPlaceholder");
+      }
+    });
+    return clone.arguments;
+  }
+
+  /**
+   * Makes an array of declarator for our VariableDeclaration
+   * @param {Array<Expression>} inits
+   * @param {Object} scope
+   */
+  function argsToVarDeclarator(inits, scope) {
+    let declarator = [];
+    declarator = inits.map(expr =>
+      t.variableDeclarator(newParamLVal(scope), expr),
+    );
+    return declarator;
+  }
+
+  /**
+   * It replaces the values of non-placeholder args in allArgs
+   * @param {Array<Declarator>} nonPlaceholderDecl that has no placeholder in them
+   * @param {Array<Arguments>} args
+   */
+  function mapNonPlaceholderToLVal(nonPlaceholderDecl, allArgsList) {
+    const clone = Array.from(allArgsList);
+    clone.forEach(cl => {
+      nonPlaceholderDecl.forEach(dec => {
+        if (dec.init.type === cl.type) {
+          if (!!cl.value && cl.value === dec.init.value) {
+            cl.value = dec.id.name;
+          } else if (!!cl.name && cl.name === dec.init.name) {
+            cl.name = dec.id.name;
+          }
+        }
+      });
+    });
+    return clone;
+  }
+
+  /**
+   * Takes the full list of arguments and extracts placeholders from it
+   * @param {Array<Argument>} allArgsList full list of arguments
+   * @returns {Array<ArgumentPlaceholder>} cloneList
+   */
+  function placeholderLVal(allArgsList) {
+    let cloneList = [];
+    allArgsList.forEach(item => {
+      if (item.name && item.name.includes("_argPlaceholder")) {
+        cloneList = cloneList.concat(item);
+      }
+    });
+    return cloneList;
+  }
+
+  return {
+    name: "proposal-partial-application",
+    inherits: syntaxPartialApplication,
+
+    visitor: {
+      CallExpression(path) {
+        if (!hasArgumentPlaceholder(path.node)) {
+          return;
+        }
+        const { node, scope } = path;
+        const functionLVal = newFuncLVal(scope);
+        const receiverLVal = newReceiverLVal(scope);
+        const nonPlaceholderArgs = unwrapArguments(node);
+        const nonPlaceholderDecl = argsToVarDeclarator(
+          nonPlaceholderArgs,
+          scope,
+        );
+        const allArgs = unwrapAllArguments(node, scope);
+        const finalArgs = mapNonPlaceholderToLVal(nonPlaceholderDecl, allArgs);
+        const placeholderVals = placeholderLVal(allArgs);
+
+        if (node.callee.type === "MemberExpression") {
+          const data = [
+            t.variableDeclaration("const", [
+              t.variableDeclarator(
+                receiverLVal,
+                t.identifier(receiverRVal(node)),
```

Comment:
You can use `path.scope.push` to generate those vars and avoid the iife:

```js
const g = o.f(?, x, 1);

// becomes

var _receiver, _func, _param, _param2; // generated by scope.push
const g = (_receiver = o, _func = o.f, _param = x, _param2 = 1, _arg => _func.call(_receiver, _arg, _param, _param2);
```

Btw, there is probably some scope/types helper function that says that `1` is a pure constant value and can thus can be transpiled as
```js
const g = (_receiver = o, _func = o.f, _param = x, _arg => _func.call(_receiver, _arg, _param, 1);
```

Reply:
I would like to ask for a bit of clarification:
- what is the advantage of not using the iife?
- could you explain how pushing to `scope` creates the declarations for me?